streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -81,12 +112,14 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         "ReceiverTracker", new ReceiverTrackerEndpoint(ssc.env.rpcEnv))
       if (!skipReceiverLaunch) receiverExecutor.start()
       logInfo("ReceiverTracker started")
+      trackerState = Started
     }
   }
 
   /** Stop the receiver execution thread. */
   def stop(graceful: Boolean): Unit = synchronized {
-    if (!receiverInputStreams.isEmpty && endpoint != null) {
+    if (isTrackerStarted) {
+      trackerState = Stopping
```

Comment:
`changing trackerState to Stopping` and `calling stopReceivers()` should be atomic, or it's still possible that `StopReceiver` won't be sent because the `registerReceiver` method runs in another thread.

And you may need to use another lock rather than adding `synchronized` to `registerReceiver` because it may block `ReceiverTrackerEndpoint`, which is actually a Akka thread. If it blocks the Akka thread for a long time, some dead-lock may happen.